README
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA1



                 ooooo     ooo   .oooooo.   ooooo
                 `888'     `8'  d8P'  `Y8b  `888'
                  888       8  888           888
                  888       8  888           888
                  888       8  888           888
                  `88.    .8'  `88b    ooo   888       o
                    `YbodP'     `Y8bood8P'  o888ooooood8


                      The UCL Compression Library
                             Version 1.01

         Copyright (C) 1996, 1997, 1998, 1999, 2000, 2001, 2002
             Markus F.X.J Oberhumer <markus@oberhumer.com>
                       http://www.oberhumer.com


 Abstract
 --------

 UCL is a portable lossless data compression library written in ANSI C.

 UCL implements a number of compression algorithms that achieve an
 excellent compression ratio while allowing *very* fast decompression.
 Decompression requires no additional memory.

 UCL is distributed under the terms of the GNU General Public License (GPL).


 Overview
 --------

 UCL implements a number of algorithms with the following features:

   - Decompression is simple and *very* fast.
   - Requires no memory for decompression.
   - The decompressors can be squeezed into less than 200 bytes of code.
   - Includes compression levels for generating pre-compressed
     data which achieve an excellent compression ratio.
   - Allows you to dial up extra compression at a speed cost in the
     compressor. The speed of the decompressor is not reduced.
   - Algorithm is thread safe.
   - Algorithm is lossless.

 UCL supports in-place decompression.


 Design criteria
 ---------------

 UCL's main design goal was a very high decompression speed while
 achieving an excellent compression ratio. Real-time decompression should
 be possible for virtually any application. The implementation of the
 NRV2B decompressor in optimized i386 assembler code runs about at
 the fifth of the speed of a memcpy() - and even faster for many files.


 Related work
 ------------

 This section describes how UCL compares to some of my other
 compression technologies.


 * LZO
   ---
   http://www.oberhumer.com/opensource/lzo/
   LZO is distributed under the terms of the GNU GPL.

   LZO is a data compression library that focuses on
   *extremly* fast decompression, and also implements some
   pretty fast compression algorithms.


 * NRV
   ---
   http://www.oberhumer.com/nrv
   NRV is not publicly available.

   NRV started life as an experimental compression library and has since
   grown into a meta-generator for compression algorithms of almost any kind.

   NRV supports a virtually unlimited number of different algorithms by
   glueing typical data compression components. By using a combination
   of sophisticated high level abstractions and advanced information
   theory concepts it usually achieves an incredible compression ratio.


 * UCL
   ---
   http://www.oberhumer.com/opensource/ucl/
   UCL is distributed under the terms of the GNU GPL.

   UCL is a re-implementation of some concrete algorithms that have
   proven especially useful during the NRV development.

   As compared to LZO, the UCL algorithms achieve a better compression
   ratio but decompression is somewhat slower. See below.


 * UPX
   ---
   http://www.oberhumer.com/opensource/upx/
   UPX is distributed under the terms of the GNU GPL.

   UPX is a very powerful executable packer that can be configured
   to use either NRV or UCL for actual compression services. It
   currently supports the NRV2B and NRV2D algorithms.


 Portability
 -----------

 UCL's decompressors should work on any system around - they could even
 get ported to 8-bit processors such as the Z-80 or 6502.

 UCL's compressors currently require at least 32-bit integers. While
 porting them to more restricted environments (such as 16-bit DOS)
 should be possible without too much effort this is not considered
 important at this time.


 How fast is fast ?
 ------------------

 Here are some rough timings on my old Intel Pentium 133:

   memcpy():                                    ~60 MB/sec
   LZO decompression in optimized assembler:    ~20 MB/sec
   LZO decompression in C:                      ~16 MB/sec
   UCL decompression in optimized assembler:    ~13 MB/sec

 Your mileage may vary and you're encouraged to run your own benchmarks.

 Also note that UCL's C language decompressors are not optimized very
 much yet, so you should use the assembler versions whenever possible.


 Documentation (preliminary)
 ---------------------------

 Currently UCL implements 3 of NRV's algorithms, namely NRV2B,
 NRV2D and NRV2E.

 UCL is a block compressor, i.e. each memory block passed to the
 compressor will get compressed independently.

 The API of UCL is basically identical to that of LZO. Due to current
 lack of documentation you are strongly advised to download LZO
 and study its documentation and example programs first:

     http://www.oberhumer.com/opensource/lzo/

 The API of UCL is actually very simple: there's a compress() function
 that compresses a block of memory, and there's a decompress() function
 that handles decompression. That's more or less all you need to know.

 See the `examples' directory for some demo programs.

 UCL will expand non-compressible data by a little amount. I suggest
 using this formula for a worst-case expansion calculation:

     output_block_size = input_block_size + (input_block_size / 8) + 256


 COPYRIGHT
 ---------

 The UCL library is Copyright (C) 1996, 1997, 1998, 1999, 2000, 2001, 2002 by
 Markus Franz Xaver Johannes Oberhumer <markus@oberhumer.com>.

 The UCL library is distributed under the terms of the GNU General Public
 License (GPL). See the file COPYING.

 Special licenses for commercial and other applications which
 are not willing to accept the GNU General Public License
 are available by contacting the author.




-----BEGIN PGP SIGNATURE-----
Version: GnuPG v1.0.6 (GNU/Linux)
Comment: For info see http://www.gnupg.org

iD8DBQE8MbkHTWFXqwsgQ8kRAskAAJ9GieiJmBBIRX01vtsIHpv49XYa9gCgswKy
WVt77NDTZENM70cU1+HmVb0=
=dI5i
-----END PGP SIGNATURE-----


NEWS
=======================================================================
User visible changes for the UCL data compression library
=======================================================================

Changes in 1.01 (02 Jan 2002)
  * Implemented the new NRV2E algorithm - better compression for some files.
  * Slightly faster and smaller i386 assembler decompressors.
  * Much better support for cross compiling.
  * Updated the configure system to use Autoconf 2.52 and Libtool 1.4.2.

Changes in 1.00 (18 Jul 2001)
  * Updated the configure system to use the latest Autoconf,
    Automake and Libtool versions.

Changes in 0.92 (10 Nov 2000)
  * Updated the UPX decompressors.

Changes in 0.91 (19 Apr 2000)
  * Added test_overlap functions that test overlapping in-place
    decompression without actually writing any decompressed data.

Changes in 0.90 (24 Feb 2000)
  * First public release.



asm/i386/gas/n2b_d_f4.S
/* n2b_d_f4.S -- ucl_nrv2b_decompress_asm_fast_safe_8

   This file is part of the UCL data compression library.

   Copyright (C) 1996-2002 Markus Franz Xaver Johannes Oberhumer
   All Rights Reserved.

   The UCL library is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License as
   published by the Free Software Foundation; either version 2 of
   the License, or (at your option) any later version.

   The UCL library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with the UCL library; see the file COPYING.
   If not, write to the Free Software Foundation, Inc.,
   59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

   Markus F.X.J. Oberhumer
   <markus@oberhumer.com>
   http://www.oberhumer.com/ucl
*/

/***** DO NOT EDIT - GENERATED AUTOMATICALLY *****/

#include "asminit.def"
typef(_ucl_nrv2b_decompress_asm_fast_safe_8)
typef(F(ucl_nrv2b_decompress_asm_fast_safe_8))
global(_ucl_nrv2b_decompress_asm_fast_safe_8)
global(F(ucl_nrv2b_decompress_asm_fast_safe_8))
_ucl_nrv2b_decompress_asm_fast_safe_8:
F(ucl_nrv2b_decompress_asm_fast_safe_8):
db 85,87,86,83,81,82,131,236,8,137,227,252,139,115,36,139
db 123,44,137,248,139,83,48,3,2,15,130,146,1,0,0,137
db 3,137,240,3,67,40,15,130,133,1,0,0,137,67,4,131
db 205,255,49,201,235,41,144,144,144,144,144,144,144,144,144,144
db 59,116,36,4,15,131,57,1,0,0,59,60,36,15,131,80
db 1,0,0,138,6,70,136,7,71,0,219,115,19,117,225,59
db 116,36,4,15,131,26,1,0,0,138,30,70,16,219,114,208
db 184,1,0,0,0,0,219,117,15,59,116,36,4,15,131,0
db 1,0,0,138,30,70,16,219,17,192,15,136,26,1,0,0
db 0,219,115,225,117,17,59,116,36,4,15,131,227,0,0,0
db 138,30,70,16,219,115,206,61,2,0,0,1,15,135,248,0
db 0,0,131,232,3,114,27,193,224,8,59,116,36,4,15,131
db 191,0,0,0,138,6,70,131,240,255,15,132,168,0,0,0
db 137,197,0,219,117,15,59,116,36,4,15,131,163,0,0,0
db 138,30,70,16,219,17,201,0,219,117,15,59,116,36,4,15
db 131,142,0,0,0,138,30,70,16,219,17,201,117,46,65,0
db 219,117,11,59,116,36,4,115,122,138,30,70,16,219,17,201
db 15,136,141,0,0,0,0,219,115,229,117,13,59,116,36,4
db 115,97,138,30,70,16,219,115,214,131,193,2,129,253,0,243
db 255,255,131,209,1,137,250,1,202,114,104,59,20,36,119,99
db 137,250,1,234,115,100,59,84,36,44,114,94,131,253,252,118
db 15,138,2,66,136,7,71,73,117,247,233,250,254,255,255,144
db 139,2,131,194,4,137,7,131,199,4,131,233,4,119,241,1
db 207,49,201,233,225,254,255,255,59,60,36,119,38,59,116,36
db 4,118,7,184,55,255,255,255,235,5,116,3,72,176,51,43
db 124,36,44,139,84,36,48,137,58,131,196,8,90,89,91,94
db 95,93,195,184,54,255,255,255,235,229,184,53,255,255,255,235
db 222,131,200,255,235,217,144,144,144,144,144,144,144,144,144,144
size(_ucl_nrv2b_decompress_asm_fast_safe_8,.-_ucl_nrv2b_decompress_asm_fast_safe_8)
size(F(ucl_nrv2b_decompress_asm_fast_safe_8),.-F(ucl_nrv2b_decompress_asm_fast_safe_8))
